### ISSUE TITLE
Many tests rely on font-size and margin being set from useragent-style-sheet

### ISSUE THREAD

**Issue by csswg-test-issue-mover** [w3c/web-platform-tests#5484]
Originally posted as https://github.com/w3c/csswg-test/issues/1097 by @hrj on 23 Jul 2016, 04:25 UTC:

> As discussed on IRC,
> 
> Take xhtml1/border-width-005.xht for example. The margin in the top text of the file factors into the result. See screenshot:
> 
> ![css21margin](https://cloud.githubusercontent.com/assets/345879/17075737/fc9dd9ca-50ba-11e6-9ed1-a57b8b8668d8.png)
> 
> The margin of the highlighted element is in `em` units and doesn't collapse with the margin of the following element. The test assumes that this margin will be computed to 16px. Hence the test passes only when font size is 16px and the user-style-sheets sets the margin of `p` element to 1em.
> 
> I think this affects more than 81 tests. I will try to enumerate them in a follow-up comment.

**Comment by csswg-test-issue-mover** [comment]
Originally posted as https://github.com/w3c/csswg-test/issues/1097#issuecomment-235291079 by @Ms2ger on 26 Jul 2016, 14:47 UTC:

> This is expected, though I'm not sure it's written down anywhere. @plinss might know.

**Comment by csswg-test-issue-mover** [comment]
Originally posted as https://github.com/w3c/csswg-test/issues/1097#issuecomment-235308536 by @plinss on 26 Jul 2016, 15:39 UTC:

> Only tests with the 'userstyle' flag should rely on a user style sheet being set, and those that do are generally testing user style sheet behavior. There are other assumptions that are relied upon that are documented in the test suite index: http://test.csswg.org/suites/css21_dev/nightly-unstable/ (and used to be documented on the wiki, not sure if they survived elsewhere). A 16px value for the 'medium' font-size is a normal assumption.

**Comment by csswg-test-issue-mover** [comment]
Originally posted as https://github.com/w3c/csswg-test/issues/1097#issuecomment-235316084 by @hrj on 26 Jul 2016, 16:01 UTC:

> Ah, so the tests expect a medium font-size of 16px. The only problem then is the assumption that `margin-bottom` of `p` element computes to `1em`.
> 
> Either that could be documented or the tests could set the margin explicitly.

**Comment by csswg-test-issue-mover** [comment]
Originally posted as https://github.com/w3c/csswg-test/issues/1097#issuecomment-235433626 by @gsnedders on 26 Jul 2016, 23:08 UTC:

> @plinss This is the UA style sheet that's in question here, not the user style sheet.
> 
> So yes, either we need to assume the margin is 1em or fix these tests. The font-size is relied on in enough places, though I don't object to people trying to fix this.

**Comment by TalbotG** [comment]
> the tests expect a medium font-size of 16px. 

Yes. I assumed, presumed and postulated that all browsers running tests would be using and relying on browser default font size (of default font face) which is 16px in the user agent style sheet of all mainstream browsers. Is such assumption posing a problem? And why would that be a problem?
The documentation originally remained silent on this precise issue although all of the built test suites are listing this as an uncommon assumption:

> The 'medium' font-size computes to 16px.
http://test.csswg.org/suites/css21_dev/20110323/#uncommon

> The only problem then is the assumption that margin-bottom of p element computes to 1em.

The default margins of `<p>` elements in the user agent style sheet of all mainstream browsers is `1em 0em`. And so, yes, the assumption is that margin-bottom of p element computes to 1em in all mainstream browsers. Is there a problem?

When I created my own tests, I wanted tests to focus only on what is being tested (target, goal) and wanted to rely on browser defaults. That makes tests' code streamlined, compact.

**Comment by hrj** [comment]
When I raised the issue, I was trying to develop and debug a [non-mainstream browser](https://gngr.info) using this test suite. Since a new browser (new layouting engine specifically) and a new team have a ton of things to figure out, it would help immensely if assumptions are documented.

Thanks for all the tests!

**Comment by TalbotG** [comment]
Okay: assumptions should be documented.Here is my proposed list:

default font-size for unstyled body text: 'medium' which should correspond to `16px`.
`body {background-color: white; color: black; margin: 8px;}`
`h1 {font-size: 2em; margin: 0.67em 0; font-weight: bold;}`
`h2 {font-size: 1.5em; margin: 0.83em 0;}`
`h3 {font-size: 1.17em;}`
`p {margin: 1em 0em;}`
`table {border-spacing: 2px;}  /* border-collapse: separate is the initial value */ `
`td {padding: 1px;}`

The user agent style sheet of all mainstream browsers follow, use, implement the above.
The rules of the user agent style sheet of Chrome, Safari and Firefox (and I presume Edge as well here) are viewable and accessible via Development Inspector tool.

Since the arrival of writing-modes spec, vertical margins in horizontal-tb writing mode have been logicalized. Firefox uses `margin-block-start` and `margin-block-end` (which is compliant with spec) while Chrome uses `-webkit-margin-before` and `-webkit-margin-after`. So, you no longer can see `p {margin: 1em 0em;}` anymore but their new names do exactly the same as before.

**Comment by TalbotG** [comment]
The assumptions in my list that are not in 
https://github.com/w3c/web-platform-tests/tree/master/infrastructure/assumptions
are:

```
body {margin: 8px;}
p {margin: 1em 0em;}
table {border-spacing: 2px;} 
/* border-collapse: separate is the initial value */
td {padding: 1px;}
```

body margins are rarely a factor in any test. 
h1, h2 and h3 are usually not part of any tests. 
p margins, border-spacing of 2px and td padding will affect a lot of tests.

Another assumption that may (can?) affect the reference file of **a few tests** is that a **'medium' border-width is 3px**. All mainstream browsers do that.
